pkg/collector/providers/zookeeper.go

```diff
@@ -60,14 +63,83 @@ func (z *ZookeeperConfigProvider) Collect() ([]check.Config, error) {
 	if err != nil {
 		return nil, err
 	}
-
+	log.Debug("the identifiers are", identifiers)
 	for _, id := range identifiers {
 		c := z.getTemplates(id)
 		configs = append(configs, c...)
 	}
 	return configs, nil
 }
 
+// Updates the list of AD templates versions in the Agent's cache and checks the list is up to date compared to Zookeeper's data.
+func (z *ZookeeperConfigProvider) IsUpToDate() (bool, error) {
+
+	updates := 0
+	var adTempAdded bool
+
+	identifiers, err := z.getIdentifiers(z.templateDir)
+	if err != nil {
+		return false, nil
+	}
+
+	// We want to be specifically notified if a template was added or removed. Then we flush the cache.
+	if len(identifiers) != len(z.cache.Adids2Node){
+		log.Infof("list of ADTemplates was modified. Cache is being updated...")
+		z.cache.Adids2Node = make(map[string]AdIdentfier2stats)
+		adTempAdded = true
+	}
+
+	for _, identifier := range identifiers {
+		// This supposes that we keep the /datadog/check_config/ad_identifiers_id/{check_names|init_configs|instances} format.
+		gChildren, _, err := z.client.Children(identifier)
+		if err != nil {
+			return false, fmt.Errorf("couldn't get key '%s' from zookeeper: %s", identifier, err)
+		}
+		newStats := AdIdentfier2stats{Stats: make(map[string]int32)}
+		for _, gcn := range gChildren {
+			gcnPath := path.Join(identifier, gcn)
+			_, stat, err := z.client.Get(gcnPath)
+			if err != nil {
+				return false, fmt.Errorf("couldn't get key '%s' from zookeeper: %s", identifier, err)
+			}
+			// Here we get the Version as opposed to the cVersion as we process the last child
+			newStats.Stats[gcn] = stat.Version //["check_config":1]
+		}
+		if len(z.cache.Adids2Node) == 0 {
+			log.Infof("Populating cache for %v.", z.String())
+			z.cache.Adids2Node[identifier] = newStats
+			continue
+		}
+
+		value, ok := z.cache.Adids2Node[identifier]
+
+		// if the template is not in the cache, add its up to date version.
+		if !ok {
+			updates++
+			z.cache.Adids2Node[identifier] = newStats
+			continue
+		}
+
+		// check if the version of the template in the cache is outdated.
+		equal, err := sameSlice(newStats.Stats, value.Stats)
+
+		if err != nil {
+			updates++
+			log.Infof("cache update failed for %v because: ", identifier, err)
+			continue
+		}
+		if !equal {
+			updates++
+			z.cache.Adids2Node[identifier] = newStats
+		}
+		log.Infof("cache up to date for %v", identifier)
+	}
+	if adTempAdded || updates > 0{
```

Comment:
We don't need to wait till the end to return. As soon as we know that templates changed we can exit, so L89 is a good time to return.
Also the cache is slightly overkill I think. I'll DM you about it.